### ISSUE TITLE
Feature proposal: use a custom function to validate the challenge in verifyAuthenticationResponse

### ISSUE THREAD

**Issue by antogyn** [MasterKale/SimpleWebAuthn#166]
It'd be useful to provide a custom challenge validation function instead of the raw `expectedChallenge`

```ts
type ChallengeValidator = (challenge: string) => boolean;
```

Why: the challenge is the only arbitrary data that is signed with WebAuthn. Its content looks like this in my case:

```ts
base64url(JSON.stringify({
  "actualChallenge": "<big random value>",
  "arbitraryDataThatMustBeSignedForOtherPurposes": {
    "data1": "....",
  }
}))
```

Only `actualChallenge` is stored and used as a nonce, and needs to be verified

**Comment by MasterKale** [comment]
That's an atypical use of `challenge`:

> As a cryptographic protocol, Web Authentication is dependent upon randomized challenges to avoid replay attacks. Therefore, the values of both PublicKeyCredentialCreationOptions.challenge and PublicKeyCredentialRequestOptions.challenge MUST be **randomly generated** by Relying Parties in an environment they trust (e.g., on the server-side), and the returned challenge value in the client’s response MUST match what was generated.

You're right that there's nothing preventing you from using `challenge` to effectively have an authenticator sign data, but for the purposes of the API specific business requirements such as yours fall outside of the purview of this library.

That said there's nothing preventing you from using the `decodeClientDataJSON()` helper beforehand to extract `actualChallenge` out and pass it in as `expectedChallenge`:

```js
import { verifyRegistrationResponse } from '@simplewebauthn/server';
import { decodeClientDataJSON } from '@simplewebauthn/server/helpers';

const clientData = decodeClientDataJSON(response.clientDataJSON);
const challengeData = base64url.toJSONProbably(clientData.challenge)

const verification = verifyRegistrationResponse({
  ...opts,
  expectedChallenge: challengeData.actualChallenge
});
```

**Comment by MasterKale** [comment]
Oh wait no, that won't work for you because the authenticator still gives back the whole data structure bytes. But you're saying you don't want to remember the whole data structure, you want to be able to pass in a helper function to evaluate the challenge by overriding how the library evaluates it...

Let me sleep on this, I'm debating if this is a use case I want to support...

**Comment by MasterKale** [comment]
@antogyn I finally woke up, and then created #172 to add this feature. I'll probably cut a release after the weekend.

**Comment by antogyn** [comment]
Great! Our workaround was to verify it ourselves and pass the full challenge to the lib (effectively bypassing it), this will simplify our code

Thanks!

**Comment by MasterKale** [comment]
I just merged PR #172 and released it as **v4.3.0**.

...But after re-reading this ticket in preparation to close it, I don't think this is done because I only implemented it for `verifyRegistrationResponse()`. Can you confirm that your actual request was that this functionality exist in `verifyAuthenticationResponse()`? For some reason I only thought to add this to registration 🤦‍♂️

**Comment by MasterKale** [comment]
I just landed PR #176 which adds custom challenge verifier functionality to `verifyAuthenticationResponse()` as well. It's currently earmarked to go out in v4.4.0, which I'll release once I address some other improvements identified in another issue.

**Comment by antogyn** [comment]
Indeed the initial request was for `verifyAuthenticationResponse()` but it's also useful in `verifyRegistrationResponse()`  in our case 👍

**Comment by MasterKale** [comment]
@antogyn I'm happy to report that I've just published **@simplewebauthn/server@4.4.0** with support for arbitrary data signing during authentication as well. Please see [CHANGELOG.md](https://github.com/MasterKale/SimpleWebAuthn/blob/master/CHANGELOG.md) for more info.
